notify/src/windows.rs

```diff
@@ -295,19 +295,16 @@ fn start_read(rd: &ReadData, event_handler: Arc<Mutex<dyn EventHandler>>, handle
             monitor_subdir,
             flags,
             &mut 0u32 as *mut u32, // not used for async reqs
-            &mut *overlapped as *mut OVERLAPPED,
+            (&mut **overlapped) as *mut OVERLAPPED,
             Some(handle_event),
         );
 
         if ret == 0 {
             // error reading. retransmute request memory to allow drop.
-            // allow overlapped to drop by omitting forget()
+            // allow overlapped to drop by dropping ManuallyDrop
             let request: Box<ReadDirectoryRequest> = mem::transmute(request_p);
-
+            std::mem::ManuallyDrop::drop(&mut overlapped);
```

Comment:
this changes the logic - we dropped it in case ret != 0, now we're dropping it right before a releasesemaphore call on ret == 0

Reply:
Also please don't name your commit "Update windows.rs"

Reply:
> Looks like it's still wrong?

How can i fix it ?

Reply:
The `drop` call should still be in the `if ret == 0` path (and none in the `else` path), but _after_ the `ReleaseSemaphore` call, so the value is dropped at the same place it would have been previously.

Reply:
Thanks!

Reply:
I would like to add that if `ManuallyDrop` is used, then [`ManuallyDrop::into_inner`](https://doc.rust-lang.org/stable/std/mem/struct.ManuallyDrop.html#method.into_inner) should be used re-arm the `Drop` impl of `Box<OVERLAPPED>`.

Personally, I have to say that using the `mem::forget` does look like the superior solution in this case, from a clarity and therefore safety perspective. Especially since the original issue description

> If a panic!() occurs between the Box::new() function and std::mem::forget, a double free vulnerability emerges.

is wrong: Conceptually, ownership of `overlapped` is passed to `ReadDirectoryChangesW` which is why we need to disarm the `Drop` when that function indicates success. So a panic would have to occur between `ReadDirectoryChangesW` returning and `mem::forget` to open the code to a potential double free, but there is no code in between for `ret != 0`. A panic between `Box::new` and `ReadDirectoryChangesW` would not imply a double-free opportunity.